sklearn/gaussian_process/gpr.py

```diff
@@ -209,12 +217,17 @@ def obj_func(theta, eval_gradient=True):
                         "Multiple optimizer restarts (n_restarts_optimizer>0) "
                         "requires that all bounds are finite.")
                 bounds = self.kernel_.bounds
-                for iteration in range(self.n_restarts_optimizer):
+
+                def optima_iterations():
                     theta_initial = \
                         self.rng.uniform(bounds[:, 0], bounds[:, 1])
                     optima.append(
                         self._constrained_optimization(obj_func, theta_initial,
                                                        bounds))
+                Parallel(n_jobs=self.n_jobs)(delayed(optima_iterations,
+                                                     check_pickle=False)()
```

Comment:
I would be really surprised if using an inner function like this was working even with `check_pickle=False` for `n_jobs != 1`... You will need to move optima_iterations to the module scope (or class-level maybe).

Drawing random numbers in parallel processes seems a bit dangerous too. And I don't think you will be able to match the result of `n_jobs=1` with `n_jobs!=-1` this way. An alternative would be to draw all the random numbers first and then pass them to the function inside Parallel.

Optional: refactoring optima_iterations into a pure function that does not take self as argument is probably a good idea to reduce surprising behaviours.